src/Autofocus.tsx

```diff
@@ -0,0 +1,159 @@
+import * as THREE from 'three'
+import React, {
+  useRef,
+  useContext,
+  useState,
+  useEffect,
+  useCallback,
+  forwardRef,
+  useImperativeHandle,
+  RefObject,
+} from 'react'
+import { useThree, useFrame, createPortal } from '@react-three/fiber'
+import { CopyPass, DepthPickingPass, DepthOfFieldEffect } from 'postprocessing'
+import { DepthOfField, EffectComposerContext } from './index'
+import { easing } from 'maath'
+
+export type AutofocusProps = typeof DepthOfField & {
+  target?: [number, number, number]
+  mouse?: boolean
+  debug?: number
+  manual?: boolean
+  smoothTime?: number
+}
+
+export type AutofocusApi = {
+  dofRef: RefObject<DepthOfFieldEffect>
+  hitpoint: THREE.Vector3
+  update: (delta: number, updateTarget: boolean) => void
+}
+
+export const Autofocus = forwardRef<AutofocusApi, AutofocusProps>(
+  (
+    { target = undefined, mouse: followMouse = false, debug = undefined, manual = false, smoothTime = 0, ...props },
+    fref
+  ) => {
+    const dofRef = useRef<DepthOfFieldEffect>(null)
+    const hitpointRef = useRef<THREE.Mesh>(null)
+    const targetRef = useRef<THREE.Mesh>(null)
+
+    const { size, gl, scene } = useThree()
+    const { composer, camera } = useContext(EffectComposerContext)
+
+    // see: https://codesandbox.io/s/depthpickingpass-x130hg
+    const [depthPickingPass] = useState(new DepthPickingPass())
+    const [copyPass] = useState(new CopyPass())
+    useEffect(() => {
+      composer.addPass(depthPickingPass)
+      composer.addPass(copyPass)
+      return () => {
+        composer.removePass(depthPickingPass)
+        composer.removePass(copyPass)
+      }
+    }, [composer, depthPickingPass, copyPass])
+
+    const [hitpoint] = useState(new THREE.Vector3(0, 0, 0))
+
+    const [ndc] = useState(new THREE.Vector3(0, 0, 0))
+    const getHit = useCallback(
+      async (x: number, y: number) => {
+        ndc.x = x
+        ndc.y = y
+        ndc.z = await depthPickingPass.readDepth(ndc)
+        ndc.z = ndc.z * 2.0 - 1.0
+        const hit = 1 - ndc.z > 0.0000001 // it is missed if ndc.z is close to 1
+        return hit ? ndc.unproject(camera) : false
+      },
+      [ndc, depthPickingPass, camera]
+    )
+
+    const [pointer] = useState(new THREE.Vector2())
+    useEffect(() => {
+      if (!followMouse) return
+
+      async function onPointermove(e: PointerEvent) {
```

Comment:
not needed imo, useFrame has that ootb: state.pointer

Reply:
omg yes, totally missed that, it a great simplification! thank you